p2p/cuprate-p2p/src/sync_states.rs

```diff
@@ -0,0 +1,429 @@
+                Ordering::Equal => {
+                    // If the cumulative difficulty of the peers chain hasn't changed then no need to update anything.
+                    return Ok(());
+                }
+                Ordering::Greater => {
+                    // This will only happen if a peer lowers its cumulative difficulty during the connection.
+                    // This won't happen if a peer re-syncs their blockchain as then the connection would have closed.
+                    tracing::debug!(
+                        "Peer's claimed cumulative difficulty has dropped, closing connection and banning peer for: {} seconds.", SHORT_BAN.as_secs()
+                    );
+                    handle.ban_peer(SHORT_BAN);
+                    return Err("Peers cumulative difficulty dropped".into());
+                }
+                Ordering::Less => (),
+            }
+
+            // Remove the old cumulative difficulty entry for this peer
+            let old_cum_dif_peers = self.cumulative_difficulties.get_mut(old_cum_dif).unwrap();
+            old_cum_dif_peers.remove(&peer_id);
+            if old_cum_dif_peers.is_empty() {
+                // If this was the last peer remove the whole entry for this cumulative difficulty.
+                self.cumulative_difficulties.remove(old_cum_dif);
+            }
+            // update the cumulative difficulty
+            *old_cum_dif = new_cumulative_difficulty;
+        } else {
+            // The peer is new so add it the list of peers.
+            self.peers.insert(
+                peer_id,
+                (
+                    new_cumulative_difficulty,
+                    PruningSeed::decompress_p2p_rules(core_sync_data.pruning_seed)?,
+                ),
+            );
+
+            // add it to the list of peers to watch for disconnection.
+            self.closed_connections.push(PeerDisconnectFut {
+                closed_fut: handle.closed(),
+                peer_id: Some(peer_id),
+            })
+        }
+
+        self.cumulative_difficulties
+            .entry(new_cumulative_difficulty)
+            .or_default()
+            .insert(peer_id);
+
+        // If the claimed cumulative difficulty is higher than the current one in the watcher
+        // or if the peer in the watch has disconnected, update it.
+        if self.new_height_watcher.borrow().cumulative_difficulty < new_cumulative_difficulty
+            || self
+                .last_peer_in_watcher_handle
+                .as_ref()
+                .is_some_and(|handle| handle.is_closed())
+        {
+            tracing::debug!(
+                "Updating sync watcher channel with new highest seen cumulative difficulty."
+            );
+            let _ = self.new_height_watcher.send(NewSyncInfo {
+                top_hash: core_sync_data.top_id,
+                chain_height: core_sync_data.current_height,
+                cumulative_difficulty: new_cumulative_difficulty,
+            });
```

Comment:
Should this be handled? What does it mean when the `Receiver` is dropped?

Reply:
It means there are no handles left, which means the network has been dropped and we are shutting down.

I don't think we need to propagate this error to the caller.